src/common/src/util/sync.rs

```diff
@@ -0,0 +1,160 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::ops::DerefMut;
+use std::sync::Arc;
+use std::time::Duration;
+
+pub type SyncPoint = String;
+pub type Signal = String;
+
+#[derive(Clone)]
+pub enum Action {
+    WaitForSignal(WaitForSignal),
+    EmitSignal(Signal),
+}
+
+#[derive(Clone)]
+pub struct WaitForSignal {
+    pub signal: Signal,
+    pub clear_signal: bool,
+    pub timeout: Duration,
+}
+
+lazy_static::lazy_static! {
+    static ref SYNC_FACILITY: SyncFacility = {
+        SyncFacility::new()
+    };
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Wait for signal {0} timeout")]
+    WaitForSignalTimeout(String),
+}
+
+#[derive(Clone)]
+struct SyncPointInfo {
+    actions: Vec<Action>,
+    execute_times: u64,
```

Comment:
could you leave some comments about what the various fields indicate?

Reply:
Added some comments.